File: C/burbuja_binaria.c
// Programar para ordenar un arreglo con bubble sort y buscar un elemento con busqueda binaria

#include <stdio.h>

// Funcion para llenar un arreglo

void fill_array(int arr[], int size)
{
    register int i;
    int element;

    for (i = 0; i < size; i++)
    {
        printf("\nIngrese el valor del elemento %d ", i);
        scanf("%d", &element);
        arr[i] = element;
    }
    printf("\n");
}

// Funcion para imprimir un arreglo

void print_array(int arr[], int size)
{
    register int i;

    printf("\nLos elementos del arreglo son: [ ");
    
    for (i = 0; i < size; i++)
    {
        printf("%d ",arr[i]);
    }
    
    printf("]\n");
}

// Funcion para el bubble sort

void bubble_sort(int arr[], int size)
{
    register int i, j;
    printf("\nOrdenando el arreglo por bubble sort");

    for (i = 0; i < size - 1; i++)
    {
        for (j = 0; j < size - i - 1; j++)
        {
            if (arr[j] > arr[j + 1])
            {
                int aux;
                aux = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = aux;
            }
        }
    }
}

// Funcion para la binary search

int binary_search(int arr[], int size, int v)
{
    int start, middle, end;
    start = 0;
    end = size - 1;

    while (start <= end)
    {
        middle = (start + end) / 2;

        if (arr[middle] == v)
            return middle;
        else if (arr[middle < v])
            start = middle + 1;
        else
            end = middle - 1;
    }

    return -1;
}

void main()
{
    int length, pos, value, n;

    printf("\nIngrese la longitud del arreglo: ");
    scanf("%d", &n);

    int array[n];

    length = sizeof(array)/sizeof(array[0]);

    // Llamada a funcion para rellenar el arreglo

    fill_array(array, length);
    print_array(array, length);

    // Llamada a funcion para el bubble sort

    bubble_sort(array, length);
    print_array(array, length);

    // Llamada a funcion para la busqueda binaria

    printf("\nIngrese el valor del elemento a buscar: ");
    scanf("%d", &value);

    pos = binary_search(array, length, value);
    printf("\nLa posicion del elemento %d es %d\n", value, pos);
}

File: Python/burbuja_binaria.py
# Programa para ordenar la lista por bubble sort y buscar un elemento por busqueda binaria

# Funcion para rellenar una lista

def fill_list(lista):
    length = int(input("\nIngrese la longitud de la lista: "))

    for i in range(0, length):
        element = int(input(f"\nIngresa el valor del elemento {i}: "))
        lista.append(element)

# Funcion para el bubble sort

def bubble_sort(lista, size):
    print("\nOrdenando la lista por bubble sort")

    for i in range(0, size - 1):
        for j in range(0, size - i - 1):
            if lista[j] > lista[j + 1]:
                aux = lista[j]
                lista[j] = lista[j + 1]
                lista[j + 1] = aux


# Funcion para la busqueda binaria

def binary_search(lista, size, v):
    start = 0
    end = size - 1

    while start <= end:

        middle = (start + end) // 2

        if lista[middle] == v:
            return middle
        elif lista[middle] > v:
            end = middle - 1
        else:
            start = middle + 1

    return -1



# MAIN PROGRAM

int_list = []

# Llamada a funcion para rellenar la lista

fill_list(int_list)
print(f"\nLa lista es: {int_list}\n")

# Llamada a funcion para el bubble sort

length = len(int_list)
bubble_sort(int_list, length)
print(f"\nLa lista ordenada es: {int_list}\n")

# Llamada a funcion para la busqueda binaria

value = int(input("Ingrese el valor del elemento a buscar: "))

pos = binary_search(int_list, length, value)
print(f"\nLa posicion del elemento {value} es {pos}")
